akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala

```diff
@@ -751,25 +772,50 @@ private[akka] class ShardRegion(
   }
 
   def replyToRegionStatsQuery(ref: ActorRef): Unit = {
-    askAllShards[Shard.ShardStats](Shard.GetShardStats)
-      .map { shardStats =>
-        ShardRegionStats(shardStats.map {
-          case (shardId, stats) => (shardId, stats.entityCount)
-        }.toMap)
+    queryShards[ShardStats](shards, Shard.GetShardStats)
+      .map {
+        case (timeouts, shardStats) =>
+          val qm = ShardingQueries.Metadata(shards.size, timeouts, shardStats)
+          if (qm.unresponsiveNonEmpty) log.warning(qm.toString)
+
+          // TODO add ShardRegionStats field for timeouts (proto): ShardRegion.ShardRegionStats(stats, timeouts.toSet)
+          // Until then
+          val stats = shardStats.map(stats => (stats.shardId, stats.entityCount)).toMap ++ timeouts.map((_, 0))
+          ShardRegionStats(stats)
       }
       .recover {
-        case _: AskTimeoutException => ShardRegionStats(Map.empty)
+        case _: AskTimeoutException =>
+          log.error("{} shards queried but timed out after {}.", shards.size, settings.shardRegionQueryTimeout)
+          ShardRegion.ShardRegionStats(shards.keySet.map(_ -> 0).toMap)
       }
       .pipeTo(ref)
   }
 
-  def askAllShards[T: ClassTag](msg: Any): Future[Seq[(ShardId, T)]] = {
+  /**
+   * Query all or a subset of shards, e.g. unresponsive shards that initially
+   * timed out. If the number of `shards` are less than this.shards.size,
+   * this could be a retry. Returns a partitioned set of any shards that may have not replied within the
+   * timeout and shards did reply, to provide retry on only that subset.
+   *
+   * To check subset unresponsive: {{{ queryShards[T](shards.filterKeys(u.contains), T) }}}
+   */
+  def queryShards[T: ClassTag](shards: Map[ShardRegion.ShardId, ActorRef], msg: Any): Future[(Seq[ShardId], Seq[T])] = {
```

Comment:
I think a specific type would be nicer here than a tuple, then we wouldn't even need the docs `case class ShardQueryResult(responses: Seq[T], timedout: Seq[ShardId])`

Reply:
I'd thought about that as well, but also why create a new thing for an internal func?

Reply:
Zero extra runtime overhead and easier to read than a comment explaining what the tuple is?

Reply:
Added.